plugins/callbacks/executors.go

```diff
@@ -0,0 +1,154 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package callbacks
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"slices"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/server/service/history/hsm"
+	"go.temporal.io/server/service/history/queues"
+)
+
+var retryable4xxErrorTypes = []int{
+	http.StatusRequestTimeout,
+	http.StatusTooManyRequests,
+}
+
+type CanGetNexusCompletion interface {
+	GetNexusCompletion(ctx context.Context) (nexus.OperationCompletion, error)
+}
+
+// HTTPCaller is a method that can be used to invoke HTTP requests.
+type HTTPCaller func(*http.Request) (*http.Response, error)
+
+type ActiveExecutorOptions struct {
+	CallerProvider func(queues.NamespaceIDAndDestination) HTTPCaller
+}
+
+func RegisterExecutor(
+	registry *hsm.Registry,
+	options ActiveExecutorOptions,
+	config *Config,
+) error {
+	exec := activeExecutor{options: options, config: config}
+	if err := hsm.RegisterExecutor(registry, TaskTypeInvocation.ID, exec.executeInvocationTask); err != nil {
+		return err
+	}
+	return hsm.RegisterExecutor(registry, TaskTypeBackoff.ID, exec.executeBackoffTask)
+}
+
+type activeExecutor struct {
+	options ActiveExecutorOptions
+	config  *Config
+}
+
+func (e activeExecutor) executeInvocationTask(ctx context.Context, env hsm.Environment, ref hsm.Ref, task InvocationTask) error {
+	ctx, cancel := context.WithTimeout(ctx, e.config.InvocationTaskTimeout())
+	defer cancel()
+
+	url, completion, err := e.loadUrlAndCallback(ctx, env, ref)
+	if err != nil {
+		return err
+	}
+
+	request, err := nexus.NewCompletionHTTPRequest(ctx, url, completion)
+	if err != nil {
+		return queues.NewUnprocessableTaskError(fmt.Sprintf("failed to construct Nexus request: %v", err))
+	}
+
+	caller := e.options.CallerProvider(queues.NamespaceIDAndDestination{NamespaceID: ref.WorkflowKey.GetNamespaceID(), Destination: task.Destination})
+	response, callErr := caller(request)
+	if callErr == nil {
+		// Body is not read but should be discarded to keep the underlying TCP connection alive.
+		// Just in case something unexpected happens while discarding or closing the body, propagate errors to the machine.
+		if _, callErr = io.Copy(io.Discard, response.Body); callErr == nil {
+			callErr = response.Body.Close()
+		}
+	}
+
+	return e.saveResult(ctx, env, ref, response, callErr)
+}
+
+func (e activeExecutor) loadUrlAndCallback(ctx context.Context, env hsm.Environment, ref hsm.Ref) (string, nexus.OperationCompletion, error) {
+	var url string
+	var completion nexus.OperationCompletion
+	err := env.Access(ctx, ref, hsm.AccessRead, func(node *hsm.Node) error {
```

Comment:
What prevents a read access from mutating the hsm.Node struct, anything?

Reply:
We discussed this elsewhere and there's not really a performant solution in Go. We could:
- copy the data and provide it that the accessor
- hash the data before and after and compare to detect mutations

but those both have overhead. Until we discover the mythical Go borrow-checker (😞) there isn't a good and high-performance solution to this.

Reply:
I think you have a good point though, and it might be better to give you a fresh copy of the node data instead of the cached copy or find a cheaper mechanism to ensure safe access.

Reply:
If we rehydrated the data blob on every access we wouldn't have this issue. How often are we actually accessing an individual HSM?

Reply:
I'd say fairly often.

Let's merge this as is and continue discussion and iterate.